presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java

```diff
@@ -0,0 +1,104 @@
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.sql.SqlFormatter.formatSql;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.lang.String.format;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class CreateFunctionTask
+        implements DataDefinitionTask<CreateFunction>
+{
+    private final SqlParser sqlParser;
+
+    @Inject
+    public CreateFunctionTask(SqlParser sqlParser)
+    {
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    @Override
+    public String getName()
+    {
+        return "CREATE FUNCTION";
+    }
+
+    @Override
+    public String explain(CreateFunction statement, List<Expression> parameters)
+    {
+        return "CREATE FUNCTION " + statement.getFunctionName();
+    }
+
+    @Override
+    public ListenableFuture<?> execute(CreateFunction statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        Analyzer analyzer = new Analyzer(stateMachine.getSession(), metadata, sqlParser, accessControl, Optional.empty(), parameters, stateMachine.getWarningCollector());
+        analyzer.analyze(statement);
+        metadata.getFunctionManager().createFunction(createFunctionToSqlInvokedFunction(statement), statement.isReplace());
+        return immediateFuture(null);
+    }
+
+    private SqlInvokedFunction createFunctionToSqlInvokedFunction(CreateFunction statement)
+    {
+        if (statement.getFunctionName().getParts().size() != 3) {
+            throw new PrestoException(GENERIC_USER_ERROR, format("Invalid SqlFunction name: %s, require exactly 3 parts", statement.getFunctionName().getParts().size()));
+        }
+
+        QualifiedFunctionName functionName = QualifiedFunctionName.of(statement.getFunctionName().toString());
+        List<SqlParameter> parameters = statement.getParameters().stream()
+                .map(parameter -> new SqlParameter(parameter.getName().toString().toLowerCase(ENGLISH), parseTypeSignature(parameter.getType())))
+                .collect(toImmutableList());
+        TypeSignature returnType = parseTypeSignature(statement.getReturnType());
+        String description = statement.getComment().orElse("");
+        RoutineCharacteristics routineCharacteristics = new RoutineCharacteristics(
+                RoutineCharacteristics.Language.valueOf(statement.getCharacteristics().getLanguage().name()),
+                RoutineCharacteristics.Determinism.valueOf(statement.getCharacteristics().getDeterminism().name()),
+                RoutineCharacteristics.NullCallClause.valueOf(statement.getCharacteristics().getNullCallClause().name()));
+        String body = formatSql(statement.getBody(), Optional.empty());
```

Comment:
It's a little strange to see `formatSql` here. Makes more sense to call it in `SHOW CREATE FUNCTION` or something like that. Consider that we didn't really do a good job in formatting expression, the original user input might even be better in a lot of cases.

Reply:
That's the only non-hacky way to convert an AST to a String representation as far as I know.